jujugui/static/gui/src/test/test_app_renderer_extension.js

```diff
@@ -0,0 +1,143 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2013 Canonical Ltd.
+
+This program is free software: you can redistribute it and/or modify it under
+the terms of the GNU Affero General Public License version 3, as published by
+the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
+General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+describe('App Renderer Extension', function() {
+
+  var createElementStub,
+      juju,
+      Renderer,
+      renderer,
+      renderStub,
+      utils,
+      Y;
+
+  before(function(done) {
+    var requires = [
+      'base',
+      'app-renderer-extension',
+      'juju-tests-utils'
+    ];
+    Y = YUI(GlobalConfig).use(requires, function(Y) {
+      juju = Y.namespace('juju');
+      utils = Y.namespace('juju-tests.utils');
+      done();
+    });
+  });
+
+  beforeEach(function() {
+    Renderer = Y.Base.create(
+        'apprenderer', Y.Base, [juju.AppRenderer], {});
+    renderer = new Renderer();
+    renderStub = utils.makeStubMethod(ReactDOM, 'render');
+    this._cleanups.push(renderStub.reset);
+    createElementStub = utils.makeStubMethod(React, 'createElement');
+    this._cleanups.push(createElementStub.reset);
+  });
+
+  afterEach(function() {
+    renderer.destroy();
+  });
+
+  describe('_renderBreadcrumb', function() {
+    beforeEach(function() {
+      var ecs = {
+        getCurrentChangeSet: utils.makeStubFunction({})
+      };
+      var Env = Y.Base.create('env', Y.Base, [], {
+        findFacadeVersion: utils.makeStubFunction('foo')
+      }, {
+        ATTRS: {
+          ecs: {}
+        }
+      });
+      renderer.env = new Env({ecs: ecs});
+      renderer.set('sandbox', false);
+      renderer._getAuth = utils.makeStubFunction({user: 'test'});
+      renderer.set('jujuEnvUUID', 'test');
+      renderer.state = {
+        getState: utils.makeStubFunction()
+      };
+      renderer.db = {
+        environment: {
+          set: utils.makeStubFunction()
+        }
+      };
+      renderer.changeState = utils.makeStubFunction();
+      renderer.showConnectingMask = utils.makeStubFunction();
+      renderer.switchEnv = utils.makeStubFunction();
+      renderer.createSocketURL = utils.makeStubFunction();
```

Comment:
I can't imagine a time when we wouldn't want to create stubs for these so could you move all of these stubs into the renderer instance on line 46. Centralize the place where we create one master 'stub' then in the beforeEach of the child suites we can overwrite them if need be.